File: cpp/problems/leetcode/0032_longest_valid_parentheses/solution.cpp
// LeetCode 32
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> st{-1};
        int best = 0;
        for (int i = 0; i < (int)s.size(); ++i) {
            if (s[i] == '(') {
                st.push_back(i);
            } else {
                st.pop_back();
                if (st.empty())
                    st.push_back(i);
                else
                    best = max(best, i - st.back());
            }
        }
        return best;
    }
};

int main() {
    assert(Solution().longestValidParentheses("(()") == 2);
    assert(Solution().longestValidParentheses(")()())") == 4);
    assert(Solution().longestValidParentheses("") == 0);
    cout << "0032 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0032_longest_valid_parentheses/solution.py
"""LeetCode 32: Longest Valid Parentheses."""

from __future__ import annotations


class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack: list[int] = [-1]
        best = 0
        for i, ch in enumerate(s):
            if ch == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    best = max(best, i - stack[-1])
        return best


if __name__ == "__main__":
    assert Solution().longestValidParentheses("(()") == 2
    assert Solution().longestValidParentheses(")()())") == 4
    assert Solution().longestValidParentheses("") == 0
    print("0032 OK")
